keps/sig-api-machinery/NNNN-watch-list/README.md

```diff
@@ -0,0 +1,956 @@
+  <th>The watch cache contacts etcd for the most up-to-date ResourceVersion.</th>
+</tr>
+<tr>
+  <th>2a.</th>
+  <th>The watch cache starts streaming initial data. The data it already has in memory.</th>
+</tr>
+<tr>
+  <th>2b.</th>
+  <th>The watch cache waits until it has observed data up to the RV received in step 2. Streaming all new data (if any) to the reflector immediately.</th>
+</tr>
+<tr>
+  <th>2c.</th>
+  <th>The watch cache has observed the RV (from step 2) and sends a bookmark event with the given RV to the reflector.</th>
+</tr>
+<tr>
+  <th>3.</th>
+  <th>The reflector replaces its internal store with collected items, updates its internal resourceVersion to the one obtained from the bookmark event.</th>
+</tr>
+<tr>
+  <th>3a.</th>
+  <th>The reflector uses the WATCH request from step 1 for further progress notifications.</th>
+</tr>
+</table>
+
+Step 1: The reflector gets a snapshot of data from the server by passing RV=”” (= from watch cache) and setting resourceVersionMatch=ConsistentInitialAdds (= ensure freshness).
+We do that on every ListAndWatch call (init or relist). 
+Each event (ADD, UPDATE, DELETE) except the BOOKMARK event received from the server is collected.
+Passing resourceVersionMatch=ConsistentInitialAdds tells the cacher it has to guarantee that the cache is at least up to date as a LIST executed at the same time.
+
+Note: Unfortunately as of today, the watch cache is vulnerable to stale reads, see https://github.com/kubernetes/kubernetes/issues/59848 for more details.
+
+Note 2: This preserves the current behavior of relisting on every ListAndWatch call.
+
+Step 2: Right after receiving a request from the reflector, the cacher gets the current resourceVersion (aka bookmarkAfterResourceVersion) directly from the etcd.
+It is used to make sure the cacher is up to date (has seen data stored in etcd) and to let the reflector know it has seen all initial data.
+There are ways to do that cheaply, e.g. we could issue a count request against the datastore.
+Next, the cacher creates a new cacheWatcher (implements watch.Interface) passing the given bookmarkAfterResourceVersion, and gets initial data from the watchCache.
+After sending initial data the cacheWatcher starts listening on an incoming channel for new events, including a bookmark event.
+At some point, the cacher will receive an event with the resourceVersion equal or greater to the bookmarkAfterResourceVersion.
+It will be propagated to the cacheWatcher and then back to the reflector as a BOOKMARK event.
+
+![watch cache with the proposed changes](./cacher.png "watch cache with the proposed changes")
+
+Step 2a: Where does the initial data come from?
+
+During construction, the cacher creates the reflector and the watchCache.
+Since the watchCache implements the Store interface it is used by the reflector to store all data it has received from etcd.
+Note: Right now the reflector used to populate the watchCache uses the old semantics of LIST/WATCH.
```

Comment:
I would just remove this - changing that also requires modifying a bit also our communication protocol with etcd, etc.
Given it's out of scope for this proposal, let's maybe remove it.

Reply:
done.